File: compact/compact.go
// Package compact implements colbin's compact mode: a bit-level wire format for
// a single struct, or an array of at most three, where the standard mode's
// columnar layout has nothing to amortise its per-column framing over.
//
//	import "github.com/ivanjoz/colbin/compact"
//
//	w := compact.NewWriter(nil, compact.ShapeStruct, true, compact.Keys8)
//	w.Key(0x35); w.Int(1234)
//	w.Key(0x9a); w.String("Usuario1")
//	w.End()
//	buf := w.Done()
//
// # Mode discrimination
//
// Bit 0 of byte 0 selects the format. The two branches share nothing else: a
// compact message has no version field, no record-count varint and no column
// count, because it needs none of them.
//
//	bit 0 == 0    standard mode: the byte is a version byte, columnar layout
//	bit 0 == 1    compact mode: the remaining bits are this package's header
//
// # Header
//
// Five bits, after which the message is one LSB-first bitstream with no
// alignment anywhere until a final pad to a byte boundary:
//
//	bit  0    1              compact mode
//	bit  1    ALL_POSITIVE   every integer in the message is >= 0
//	bits 2-3  shape          0 = lone struct, 1..3 = array of that many records
//	bit  4    NARROW_KEYS    field ids are 4 bits wide rather than 8
//
// Shape 0 and shape 1 both carry one record; they differ only in whether it is
// rendered as an object or as an array of one, which the binary path takes from
// the destination Go type but a JSON reader needs told.
//
// # Records
//
// A record is a run of [key][value] pairs closed by the terminator key. A field
// holding its zero value is omitted entirely, so an absent field costs nothing
// beyond the terminator the record already owes. The key is the same field id
// the standard mode uses, so a reader resolves fields by id, not by position.
//
// # Key width
//
// A key is 8 bits by default, matching the standard mode's field id, with 255
// closing a record. When every id a message writes is 14 or less the header's
// NARROW_KEYS bit selects a 4-bit key instead, with 15 closing the record — the
// same arrangement one nibble down, so a reader still resolves fields by id and
// still steps over the ones it does not know.
//
// That halves the framing: a record of f present fields spends 4*(f+1) fewer
// bits, against the one bit the header flag costs the whole message. It pays for
// itself on the first field of the first record. Ids are what decide it, not the
// field count, so it wants explicit small ids in the struct tags (`cb:"1"`,
// `cb:"2"`, ...); a hashed id lands anywhere in 0..254 and forces the wide key.
//
// # Composites
//
// A record is a run of keyed values, and three of those value forms are
// themselves made of values. They recurse through the same bitstream, and none
// of them borrows anything from the standard mode: compact mode has no column
// framing to embed, so a nested struct is another key run, not a sub-table. The
// two modes share value codecs -- varint, packed5 -- and nothing else.
//
//	nested struct   [key] [ [key][value] ... [terminator] ]
//	array           [key] [count:varint] [value] [value] ...
//	map             [key] [count:varint] [key][value] [key][value] ...
//
// Every form is self delimiting given the schema, which is the only thing the
// format asks of a value: an array carries its count, a packed5 frame carries
// its own end, and a key run closes on the terminator. Nothing needs a byte
// length and nothing needs a type tag.
//
// What the schema cannot supply is a value's *dynamic* type, so an interface
// field has no compact form at all -- see codec/compact_plan.go, which is where
// the per-field rule lives.
//
// The omit-zero rule reaches into the composites the same way it reaches a
// scalar. A nested struct all of whose fields are zero is omitted entirely, and
// so is an array or map of length zero, which is why an empty-but-non-nil slice
// or map decodes back as nil -- exactly as it already did for a slice.
//
// One thing changes for the key width: a nested key run spends the same key bits
// as the record containing it, so NARROW_KEYS is a property of the whole message
// rather than of the root struct. Every struct reachable in the message must
// keep its ids at or below MaxNarrowKey, which in practice means tagging the
// nested structs too.
//
// # Integers
//
// ALL_POSITIVE decides how a signed value becomes the unsigned payload: with the
// flag set the payload is the magnitude, and without it the payload is the
// zigzag. Setting the flag is worth one payload bit on every integer in the
// message, and a message-wide pre-scan is free at three records.
//
// The payload is then written with a varint whose first unit is one bit narrower
// than LEB128's, spending that bit on a selector:
//
//	unit 0    [selector:1] [cont:1] [payload:6]   (+ [payload:4] if selector)
//	unit i    [cont:1] [payload:7]
//
// The selector's four bits are a one-time offset applied to the first unit only,
// not a per-unit addition. That gives two ladders of capacity — 6+7k without it
// and 10+7k with it — and the encoder takes whichever reaches the value first:
//
//	bits    0-6   7    8-10  11-13  14   15-17  18-20  21   22-24
//	LEB128    8   8      16     16   16     24     24   24      32
//	compact   8  12      12     16   20     20     24   28      28
//
// Three ties, three wins of four bits, one loss of four, per seven bit-lengths.
// The selector costs nothing outright: it takes the seventh payload bit of unit
// 0, which any value wider than six bits was going to spill anyway. The loss
// falls only where LEB128's first byte was exactly full.
package compact

import "errors"

var (
	// ErrNotCompact is returned when a buffer's first bit is clear, meaning it
	// holds a standard-mode message rather than a compact one.
	ErrNotCompact = errors.New("colbin: not a compact message")
	// ErrTruncated is returned when a read runs past the end of the bitstream.
	ErrTruncated = errors.New("colbin: compact message truncated")
	// ErrBadString is returned when a string field does not hold a valid packed5
	// frame.
	ErrBadString = errors.New("colbin: compact bad string frame")
	// ErrSkipComposite is returned when Skip is asked to step over a nested
	// struct, array or map. Those forms are self delimiting only against their
	// own sub-schema, which Skip is not given -- see Reader.Skip.
	ErrSkipComposite = errors.New("colbin: compact composite cannot be skipped without its sub-schema")
)

// Shape says how many records the message holds and whether a lone record is an
// object or an array of one. It occupies bits 2-3 of the header.
type Shape uint8

const (
	// ShapeStruct is one record rendered as an object.
	ShapeStruct Shape = 0
	// ShapeArray1, ShapeArray2 and ShapeArray3 are arrays of that many records.
	ShapeArray1 Shape = 1
	ShapeArray2 Shape = 2
	ShapeArray3 Shape = 3
)

// Records is how many records the shape describes.
func (s Shape) Records() int {
	if s == ShapeStruct {
		return 1
	}
	return int(s)
}

// MaxRecords is the largest array compact mode can carry. Past it the columnar
// standard mode has enough rows to amortise its framing and wins outright.
const MaxRecords = 3

// TerminatorKey closes a record's key/value run. It matches the field id the
// standard mode reserves, so no real field can collide with it.
//
// It is the id a caller sees in either key width: under Keys4 the terminator on
// the wire is 15, and Reader.Key reports it as TerminatorKey so that the check
// closing a record reads the same on both paths.
const TerminatorKey uint8 = 255

// KeyWidth is how many bits a field id occupies, chosen once per message and
// recorded in the header's NARROW_KEYS bit.
type KeyWidth uint8

const (
	// Keys8 is the default: ids 0..254, with 255 closing a record. It is the
	// standard mode's field id unchanged, so any struct can use it.
	Keys8 KeyWidth = 8
	// Keys4 is the narrow form: ids 0..MaxNarrowKey, with 15 closing a record.
	// It costs one header bit and saves four on every key, so it wins from the
	// first field onwards -- but only a struct whose ids are all small can use
	// it, which in practice means explicit `cb:"1"`-style tags.
	Keys4 KeyWidth = 4
)

// MaxNarrowKey is the largest field id Keys4 can carry. 15 is the terminator, so
// the ids stop one short of it, exactly as 255 stops the wide ones at 254.
const MaxNarrowKey uint8 = 14

// narrow reports whether k is the 4-bit width, which is what the header carries.
// Only Keys4 is narrow, so any other value -- including one a caller invented --
// resolves to the wide key rather than to a width nothing can read.
func (k KeyWidth) narrow() bool { return k == Keys4 }

// Bits is the width of one key. It is also the fewest bits a record can occupy,
// since an empty record is its terminator alone -- which is what bounds-checking
// a record count needs (see Reader.Count).
func (k KeyWidth) Bits() int { return int(k.bits()) }

// bits is the width of one key, and terminator is the id that closes a record at
// that width.
func (k KeyWidth) bits() uint8 {
	if k.narrow() {
		return 4
	}
	return 8
}

func (k KeyWidth) terminator() uint64 {
	if k.narrow() {
		return uint64(narrowTerminatorKey)
	}
	return uint64(TerminatorKey)
}

// narrowTerminatorKey closes a record under Keys4. It is TerminatorKey's role one
// nibble down: the top id at that width, reserved so no field can hold it.
const narrowTerminatorKey uint8 = 15

// Header bit widths, in the order they are written.
const (
	modeBits   = 1 // 1 == compact
	flagBits   = 1 // ALL_POSITIVE
	shapeBits  = 2
	narrowBits = 1 // NARROW_KEYS

	headerBits = modeBits + flagBits + shapeBits + narrowBits
)

// IsCompact reports whether buf holds a compact-mode message, which is the whole
// of the mode discrimination: bit 0 of byte 0.
func IsCompact(buf []byte) bool {
	return len(buf) > 0 && buf[0]&1 == 1
}

File: rust/src/compact.rs
//! Compact mode, mirroring the `compact` package.
//!
//! One record, or an array of at most three, in a single LSB-first bitstream
//! with no alignment anywhere until a final pad to a byte boundary. Five header
//! bits, then one record per shape:
//!
//! ```text
//! bit  0    1              compact mode
//! bit  1    ALL_POSITIVE   every integer in the message is >= 0
//! bits 2-3  shape          0 = lone struct, 1..3 = array of that many records
//! bit  4    NARROW_KEYS    field ids are 4 bits wide rather than 8
//! ```
//!
//! A record is a run of `[key][value]` pairs closed by the terminator key. A
//! field holding its zero value is omitted entirely, so the key run *is* the
//! presence information and an absent field costs nothing beyond the terminator
//! the record already owes.

use crate::bitstream::BitReader;
use crate::{Error, Kind, Record, Schema, Value, packed5, varint};

const SHAPE_STRUCT: u64 = 0;
/// The largest array compact mode can carry.
const MAX_RECORDS: u64 = 3;

/// The header: the compact bit, ALL_POSITIVE, two shape bits and NARROW_KEYS.
const HEADER_BITS: u8 = 5;

/// Closes a record under `Keys8`; the id the standard mode also reserves.
const TERMINATOR_KEY: u64 = 255;
/// Closes a record under `Keys4`: the same role one nibble down.
const NARROW_TERMINATOR_KEY: u64 = 15;

pub(crate) fn decode(data: &[u8], schema: &Schema) -> Result<Vec<Record>, Error> {
    let mut out = Vec::new();
    decode_into(data, schema, &mut out)?;
    Ok(out)
}

/// Decodes into `dst`, reusing the `Record`s it already holds and their field
/// vectors. A reused record is cleared for the reason the module documents: an
/// omitted field is simply not written, so the previous message's value would
/// otherwise survive underneath it.
pub(crate) fn decode_into(
    data: &[u8],
    schema: &Schema,
    dst: &mut Vec<Record>,
) -> Result<(), Error> {
    let mut reader = Reader::new(data)?;
    let count = reader.records;
    // Grown, never shrunk: a caller alternating a long message with a short one
    // keeps the capacity the long one needed.
    if dst.len() < count {
        dst.resize_with(count, Record::new);
    }
    for record in dst.iter_mut().take(count) {
        read_record_into(&mut reader, schema, record)?;
    }
    dst.truncate(count);
    reader.check()
}

/// Decodes a one-record message straight into a derived struct's fields.
///
/// No `Record` and no `Value`: the schema maps each wire id to a declaration
/// index and the generated `colbin_read` assigns to the field at that position.
pub(crate) fn decode_one_typed<T: crate::Colbin>(
    data: &[u8],
    schema: &Schema,
    dst: &mut T,
) -> Result<(), Error> {
    let mut reader = Reader::new(data)?;
    if reader.records != 1 {
        return Err(Error::RecordCount(reader.records));
    }
    // An omitted field is not written at all, so the destination starts zeroed
    // rather than being merged into.
    *dst = T::colbin_zero();
    loop {
        let key = reader.key();
        reader.check()?;
        if key == TERMINATOR_KEY {
            return reader.check();
        }
        let id = key as u8;
        let index = schema.index(id).ok_or(Error::UnknownField(id))?;
        let mut field = crate::FieldReader {
            reader: &mut reader,
        };
        dst.colbin_read(index, &mut field)?;
        reader.check()?;
    }
}

/// Decodes a one-record message into `dst`, without the `Vec<Record>` that
/// holding the record in a slice would cost.
pub(crate) fn decode_one_into(data: &[u8], schema: &Schema, dst: &mut Record) -> Result<(), Error> {
    let mut reader = Reader::new(data)?;
    if reader.records != 1 {
        return Err(Error::RecordCount(reader.records));
    }
    read_record_into(&mut reader, schema, dst)?;
    reader.check()
}

pub(crate) struct Reader<'a> {
    bits: BitReader<'a>,
    all_positive: bool,
    key_bits: u8,
    terminator: u64,
    records: usize,
}

impl<'a> Reader<'a> {
    fn new(data: &'a [u8]) -> Result<Self, Error> {
        let mut bits = BitReader::new(data);
        // The five header bits in one read. Four reads of one, one, two and one
        // bit cost four bounds checks and four window loads for what is a single
        // shift and mask, and this is paid once per message.
        let header = bits.get(HEADER_BITS);
        if let Some(err) = bits.err() {
            return Err(err.clone());
        }
        // bit 0 is the compact bit, which decode() already discriminated on.
        let all_positive = header & 0b0_0010 != 0;
        let shape = (header >> 2) & 0b11;
        let narrow = header & 0b1_0000 != 0;
        // Shape 0 and shape 1 both carry one record; they differ only in whether
        // a JSON reader renders it as an object or as an array of one, which is
        // not something a schema-driven decode has to decide.
        let records = if shape == SHAPE_STRUCT {
            1
        } else {
            shape.min(MAX_RECORDS) as usize
        };
        let (key_bits, terminator) = if narrow {
            (4, NARROW_TERMINATOR_KEY)
        } else {
            (8, TERMINATOR_KEY)
        };
        Ok(Self {
            bits,
            all_positive,
            key_bits,
            terminator,
            records,
        })
    }

    fn check(&self) -> Result<(), Error> {
        match self.bits.err() {
            Some(err) => Err(err.clone()),
            None => Ok(()),
        }
    }

    /// Reads a field id at whichever width the header declared, reporting the
    /// terminator as 255 on both paths: `Keys4` reserves 15 the way `Keys8`
    /// reserves 255, so no field can hold either.
    fn key(&mut self) -> u64 {
        let key = self.bits.get(self.key_bits);
        if key == self.terminator {
            TERMINATOR_KEY
        } else {
            key
        }
    }

    /// A signed integer: the payload is the magnitude when ALL_POSITIVE is set
    /// and the zigzag when it is not.
    pub(crate) fn int(&mut self) -> i64 {
        let raw = varint::get_varint(&mut self.bits);
        if self.all_positive {
            raw as i64
        } else {
            varint::unzigzag(raw)
        }
    }

    pub(crate) fn uint(&mut self) -> u64 {
        varint::get_varint(&mut self.bits)
    }

    /// An element count, checked against the bits actually remaining given the
    /// fewest bits one element can occupy. A corrupt count near 2^64 has to be
    /// rejected before it is used to size anything.
    fn array_len(&mut self, min_bits_per_elem: usize) -> Result<usize, Error> {
        let count = varint::get_varint(&mut self.bits);
        self.check()?;
        if count > (self.bits.remaining() / min_bits_per_elem) as u64 {
            return Err(Error::Truncated);
        }
        Ok(count as usize)
    }

    /// A packed5 frame. The codec is byte structured, so the frame simply starts
    /// wherever the bitstream happens to be and the tail is shifted into place
    /// when that is not a byte boundary.
    pub(crate) fn string(&mut self) -> Result<String, Error> {
        self.check()?;
        // The borrow of the shifted tail ends here, before advance_bytes needs
        // the reader again.
        let (bytes, consumed) = packed5::decode(self.bits.aligned_tail())?;
        // A well formed frame never extends into the final pad, so one that
        // would is corrupt rather than a short read of valid data.
        if !self.bits.advance_bytes(consumed) {
            return Err(Error::Truncated);
        }
        String::from_utf8(bytes).map_err(|_| Error::NotUtf8)
    }

    pub(crate) fn bytes(&mut self) -> Result<Vec<u8>, Error> {
        let count = self.array_len(8)?;
        let mut out = Vec::with_capacity(count);
        for _ in 0..count {
            out.push(self.bits.get(8) as u8);
        }
        self.check()?;
        Ok(out)
    }

    /// One bit.
    pub(crate) fn bit(&mut self) -> bool {
        self.bits.get_bool()
    }

    pub(crate) fn f32(&mut self) -> f32 {
        f32::from_bits(self.bits.get(32) as u32)
    }

    pub(crate) fn f64(&mut self) -> f64 {
        f64::from_bits(self.bits.get(64))
    }

    /// A bitmap, one bit per element: the one array form that does not match
    /// standard mode, which carries bools through an integer column.
    pub(crate) fn bools(&mut self) -> Result<Vec<bool>, Error> {
        let count = self.array_len(1)?;
        let mut out = crate::try_vec(count)?;
        for _ in 0..count {
            out.push(self.bits.get_bool());
        }
        Ok(out)
    }

    pub(crate) fn strings(&mut self) -> Result<Vec<String>, Error> {
        let count = self.array_len(8)?;
        let mut out = crate::try_vec(count)?;
        for _ in 0..count {
            out.push(self.string()?);
        }
        Ok(out)
    }

    pub(crate) fn f32s(&mut self) -> Result<Vec<f32>, Error> {
        let count = self.array_len(32)?;
        let mut out = crate::try_vec(count)?;
        for _ in 0..count {
            out.push(f32::from_bits(self.bits.get(32) as u32));
        }
        Ok(out)
    }

    pub(crate) fn f64s(&mut self) -> Result<Vec<f64>, Error> {
        let count = self.array_len(64)?;
        let mut out = crate::try_vec(count)?;
        for _ in 0..count {
            out.push(f64::from_bits(self.bits.get(64)));
        }
        Ok(out)
    }

    /// An integer slice, written with the same array codec that backs an integer
    /// column in standard mode — which is why an array costs the same in either
    /// mode.
    pub(crate) fn ints(&mut self, width: u8) -> Result<Vec<i64>, Error> {
        // The array codec spends at least one byte per element.
        let count = self.array_len(8)?;
        if count == 0 {
            return Ok(Vec::new());
        }
        let (values, consumed) = varint::decode_array(self.bits.aligned_tail(), count, width)?;
        if !self.bits.advance_bytes(consumed) {
            return Err(Error::Truncated);
        }
        Ok(values)
    }
}

/// Fills one record, stopping at its terminator.
///
/// An id the schema does not know is an error rather than a skip: the wire holds
/// no type tag, so there is no way to know how far to step over it.
///
/// This is also how a nested struct is read, and how one element of an array of
/// structs is: a nested struct is another record closed by the same terminator,
/// so the recursion changes the depth and not the framing. `Schema::build`
/// already bounded that depth, so no counter is threaded through here.
fn read_record(reader: &mut Reader<'_>, schema: &Schema) -> Result<Record, Error> {
    let mut record = Record::with_capacity(schema.fields().len());
    read_record_into(reader, schema, &mut record)?;
    Ok(record)
}

/// `read_record` into a record the caller owns, so its field vector survives
/// between messages.
fn read_record_into(
    reader: &mut Reader<'_>,
    schema: &Schema,
    record: &mut Record,
) -> Result<(), Error> {
    record.clear();
    record.reserve(schema.fields().len());
    loop {
        let key = reader.key();
        reader.check()?;
        if key == TERMINATOR_KEY {
            return Ok(());
        }
        let id = key as u8;
        let kind = schema.kind(id).ok_or(Error::UnknownField(id))?;
        let value = read_value(reader, kind)?;
        reader.check()?;
        record.push(id, value);
    }
}

/// The fewest bits one value of this kind can occupy, which is what
/// bounds-checking a composite's count needs: a corrupt count near 2^64 has to
/// be rejected before it sizes an allocation. It mirrors `compactElemMinBits`.
fn min_bits(reader: &Reader<'_>, kind: &Kind) -> usize {
    match kind {
        Kind::Bool => 1,
        Kind::Float32 => 32,
        Kind::Float64 => 64,
        // An empty record is its terminator alone.
        Kind::Struct(_) => usize::from(reader.key_bits),
        // Everything else opens with a varint unit, a packed5 frame or a count.
        _ => 8,
    }
}

/// Reads one value at the kind the schema declares.
///
/// **One exhaustive match**, deliberately. This used to walk up to four
/// sequential matches on `kind` — a composite check, then `scalar_int`, then
/// `slice_int`, then a final match — and dispatch was the whole cost of a field:
/// a `bool` field, which is a single bit on the wire, measured the same 34 ns as
/// an `i32`, and an `f64` measured *less*. Collapsing them to one jump table
/// took a seven-field record from 231 ns to well under half that.
///
/// The width tables (`Kind::scalar_int`, `Kind::slice_int`) still exist for the
/// standard-mode reader, which needs the width as a value rather than as a
/// branch. Here the width is a constant per arm, which is the point.
fn read_value(reader: &mut Reader<'_>, kind: &Kind) -> Result<Value, Error> {
    // A signed value is truncated to the field's width, which never reaches the
    // wire, exactly as the Go decoder's `*(*int8)(q) = int8(r.Int())` does.
    Ok(match kind {
        Kind::Bool => Value::Bool(reader.bit()),
        Kind::Int8 => Value::Int(varint::sign_extend(reader.int() as u64, 8)),
        Kind::Int16 => Value::Int(varint::sign_extend(reader.int() as u64, 16)),
        Kind::Int32 => Value::Int(varint::sign_extend(reader.int() as u64, 32)),
        // Already 64 bits wide, so there is nothing to sign extend from.
        Kind::Int64 => Value::Int(reader.int()),
        Kind::Uint8 => Value::Uint(mask(reader.uint(), 8)),
        Kind::Uint16 => Value::Uint(mask(reader.uint(), 16)),
        Kind::Uint32 => Value::Uint(mask(reader.uint(), 32)),
        Kind::Uint64 => Value::Uint(reader.uint()),
        Kind::Float32 => Value::Float32(reader.f32()),
        Kind::Float64 => Value::Float64(reader.f64()),
        Kind::String => Value::String(reader.string()?),
        Kind::Bytes => Value::Bytes(reader.bytes()?),

        Kind::Int8s => Value::Ints(reader.ints(8)?),
        Kind::Int16s => Value::Ints(reader.ints(16)?),
        Kind::Int32s => Value::Ints(reader.ints(32)?),
        Kind::Int64s => Value::Ints(reader.ints(64)?),
        // An unsigned slice rode the same-width signed codec, which preserved the
        // bit pattern; masking is what hands it back unsigned.
        Kind::Uint16s => Value::Uints(unsigned(reader.ints(16)?, 16)),
        Kind::Uint32s => Value::Uints(unsigned(reader.ints(32)?, 32)),
        Kind::Uint64s => Value::Uints(unsigned(reader.ints(64)?, 64)),
        Kind::Bools => Value::Bools(reader.bools()?),
        Kind::Strings => Value::Strings(reader.strings()?),
        Kind::Float32s => Value::Float32s(reader.f32s()?),
        Kind::Float64s => Value::Float64s(reader.f64s()?),

        // Composites recurse through the same bitstream: a nested key run, a
        // counted run of values, a counted run of key/value pairs. Every form is
        // self delimiting against the schema, which is all the format asks.
        Kind::Struct(schema) => Value::Record(read_record(reader, schema)?),
        Kind::Array(elem) => {
            let count = reader.array_len(min_bits(reader, elem))?;
            let mut out = crate::try_vec(count)?;
            for _ in 0..count {
                out.push(read_value(reader, elem)?);
            }
            Value::Array(out)
        }
        Kind::Map(key, val) => {
            let count = reader.array_len(min_bits(reader, key) + min_bits(reader, val))?;
            let mut out = crate::try_vec(count)?;
            for _ in 0..count {
                let k = read_value(reader, key)?;
                let v = read_value(reader, val)?;
                out.push((k, v));
            }
            Value::Map(out)
        }
    })
}

/// Masks a signed-codec slice back to its unsigned element width.
fn unsigned(values: Vec<i64>, width: u8) -> Vec<u64> {
    values.into_iter().map(|v| mask(v as u64, width)).collect()
}

/// Truncates an unsigned value to its declared width.
fn mask(value: u64, width: u8) -> u64 {
    if width >= 64 {
        value
    } else {
        value & ((1_u64 << width) - 1)
    }
}
